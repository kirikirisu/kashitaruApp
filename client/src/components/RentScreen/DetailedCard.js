import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import clsx from 'clsx';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardMedia from '@material-ui/core/CardMedia';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import Collapse from '@material-ui/core/Collapse';
import Avatar from '@material-ui/core/Avatar';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
// import FavoriteIcon from '@material-ui/icons/Favorite';
import Chat from '@material-ui/icons/Chat';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import MoreVertIcon from '@material-ui/icons/MoreVert';
import connectToRoom from '../../utils/connectToRoom';

const useStyles = makeStyles((theme) => ({
  card: {
    maxWidth: 345,
  },
  media: {
    height: 0,
    paddingTop: '56.25%', // 16:9
  },
  expand: {
    transform: 'rotate(0deg)',
    marginLeft: 'auto',
    transition: theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    }),
  },
  expandOpen: {
    transform: 'rotate(180deg)',
  },
  avatar: {
    margin: 10,
  },
}));

const DetailedCard = ({
  productName,
  img,
  description,
  price,
  period,
  shippingArea,
  days,
  name,
  avatar,
  comment,
  currentUser,
  rooms,
  redirectChat,
  toggleRedirectChat,
  rest,
}) => {
  const classes = useStyles();
  const [expanded, setExpanded] = React.useState(false);

  const handleExpandClick = () => {
    setExpanded(!expanded);
  };

  const createPrivateRoom = (id) => {
    const roomName = `${currentUser.id}_${id}`;

    const isPrivateChatCreated = rooms.filter((room) => {
      if (room.customData && room.customData.isDirectMessage) {
        const arr = [currentUser.id, id];
        const { userIds } = room.customData;

        if (arr.sort().join('') === userIds.sort().join('')) {
          return {
            room,
          };
        }
      }

      return false;
    });

    if (isPrivateChatCreated.length > 0) {
      return Promise.resolve(isPrivateChatCreated[0]);
    }

    return currentUser.createRoom({
      name: `${roomName}`,
      private: true,
      addUserIds: [`${id}`],
      customData: {
        isDirectMessage: true,
        userIds: [currentUser.id, id],
      },
    });
  };

  const sendDM = (id) => {
    createPrivateRoom(id).then((room) => {
      toggleRedirectChat(true);
      connectToRoom(room.id, currentUser, rest);
    });
  };

  const renderDetailedCard = () => (
    <Card className={classes.card}>
      <CardHeader
        avatar={
          <Avatar aria-label="recipe" className={classes.avatar} src={avatar} />
        }
        action={(
          <IconButton aria-label="settings">
            <MoreVertIcon />
          </IconButton>
        )}
        title={name}
        subheader="投稿日2019_3_9"
      />
      <CardMedia
        className={classes.media}
        image={img}
        title="product image"
      />
      <CardContent>
        <Typography variant="h6" color="textSecondary" component="p">
          {productName}
        </Typography>
      </CardContent>
      <CardActions disableSpacing>
        <IconButton aria-label="chat" onClick={() => sendDM(name)}>
          <Chat />
        </IconButton>
        <IconButton
          className={clsx(classes.expand, {
            [classes.expandOpen]: expanded,
          })}
          onClick={handleExpandClick}
          aria-expanded={expanded}
          aria-label="show more"
        >
          <ExpandMoreIcon />
        </IconButton>
      </CardActions>
      <Collapse in={expanded} timeout="auto" unmountOnExit>
        <CardContent>
          <Typography paragraph>商品説明:</Typography>
          <Typography paragraph>
            {description}
          </Typography>
          <Typography paragraph>価格:</Typography>
          <Typography paragraph>
            {price}
          </Typography>
          <Typography paragraph>貸し出し期間:</Typography>
          <Typography paragraph>
            {period}
          </Typography>
          <Typography paragraph>発送元エリア:</Typography>
          <Typography paragraph>
            {shippingArea}
          </Typography>
          <Typography paragraph>発送日数:</Typography>
          <Typography paragraph>
            {days}
          </Typography>
        </CardContent>
      </Collapse>
    </Card>
  );

  return (
    <Route
      render={() => (
        redirectChat ? (
          <Redirect to="chat" />
        ) : (
            renderDetailedCard()
          )
      )}
    />
  );
};

export default DetailedCard;
