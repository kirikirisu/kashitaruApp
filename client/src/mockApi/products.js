const products = [
  {
    "id": 7,
    "productImg": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/Android.png?alt=media&token=aae1f59d-0da8-434a-bb0c-91808835f2e4",
    "productName": "カッパの尻子玉",
    "description": "村長のやつです",
    "price": "９万円",
    "period": "８ヶ月",
    "shippingArea": "荒川区",
    "days": "３日",
    "name": "カッパ",
    "avatar": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/20140321122433e1d.jpg?alt=media&token=86086470-f12c-4cb6-869e-9ff8416cd887",
    "comment": "はい妖怪幻術甲羅はずれてるように見えたの術"
  },
  {
    "id": 7,
    "productImg": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/Go.png?alt=media&token=ff2fe281-1403-4f1b-9584-5501cc388e51",
    "productName": "カッパの皿",
    "description": "村長のやつ",
    "price": "３千円",
    "period": "９ヶ月",
    "shippingArea": "荒川区",
    "days": "９日",
    "name": "カッパ",
    "avatar": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/20140321122433e1d.jpg?alt=media&token=86086470-f12c-4cb6-869e-9ff8416cd887",
    "comment": "はい妖怪幻術甲羅はずれてるように見えたの術"
  },
  {
    "id": 7,
    "productImg": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/rails2-logo.png?alt=media&token=589c6622-e31e-4499-917b-5af119b55277",
    "productName": "fsdfsdf",
    "description": "fsdfsdfsdf",
    "price": "fdsf",
    "period": "fdsff",
    "shippingArea": "fdsf",
    "days": "fsdf",
    "name": "カッパ",
    "avatar": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/20140321122433e1d.jpg?alt=media&token=86086470-f12c-4cb6-869e-9ff8416cd887",
    "comment": "はい妖怪幻術甲羅はずれてるように見えたの術"
  },
  {
    "id": 7,
    "productImg": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/docker.png?alt=media&token=7fe28d4a-f2de-49e2-be42-b05914f5b692",
    "productName": "jhkhkhj",
    "description": "kkm",
    "price": "hbth",
    "period": "thth",
    "shippingArea": "gerg",
    "days": "gerg",
    "name": "カッパ",
    "avatar": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/20140321122433e1d.jpg?alt=media&token=86086470-f12c-4cb6-869e-9ff8416cd887",
    "comment": "はい妖怪幻術甲羅はずれてるように見えたの術"
  },
  {
    "id": 8,
    "productImg": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/Gitlab.png?alt=media&token=b9d44c7b-5467-4956-a99e-b38c21f3b0ec",
    "productName": "星の被り物",
    "description": "黄色いやつです",
    "price": "無料",
    "period": "５ヶ月",
    "shippingArea": "荒川区",
    "days": "２日",
    "name": "ほし",
    "avatar": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/7c47b5da.jpg?alt=media&token=c5ca5944-f442-40d5-8587-01ca148c9f17",
    "comment": "「水着の女の子のいるところ」...それだけで十分なんだよ！！"
  },
  {
    "id": 9,
    "productImg": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/VScode.png?alt=media&token=e3a51472-f0f3-4234-8d22-e6975bc68ea9",
    "productName": "hogehoge",
    "description": "hogehogehogehogeghoegjeogheogheog",
    "price": "gwegwegwgwg",
    "period": "gwregrg",
    "shippingArea": "gererg",
    "days": "gregerg",
    "name": "nino",
    "avatar": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/images-3-12.jpg?alt=media&token=5e6c4888-b9e6-4afb-8b82-c5c42ccd3ffc",
    "comment": "私は金星人だ"
  },
  {
    "id": 9,
    "productImg": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/React.png?alt=media&token=0a1dc8be-29a3-4fd3-b8c7-07c5bf49bbbe",
    "productName": "sfew",
    "description": "fewfwe",
    "price": "fwe",
    "period": "fw",
    "shippingArea": "fwewf",
    "days": "gweg",
    "name": "nino",
    "avatar": "https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/images-3-12.jpg?alt=media&token=5e6c4888-b9e6-4afb-8b82-c5c42ccd3ffc",
    "comment": "私は金星人だ"
  }
];

export default products;
