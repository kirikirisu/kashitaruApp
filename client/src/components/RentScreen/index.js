import React from 'react';
import Grid from '@material-ui/core/Grid';
import DetailedCard from './DetailedCard';
import key from '../../utils/listKeyGenerator';

const RentScreen = ({
  isFetching,
  shareInformationsArray,
  currentUser,
  rooms,
  toggleRedirectChat,
  redirectChat,
  ...rest // connectToRoomのアクションたち
}) => (
    <div>
      {
        isFetching // isFetchingの値で分岐
          ? <h2>Now Loading...</h2> // データをFetch中ならばローディングアイコンを表示
          : (
            <div>
              <Grid item xs={12}>
                <Grid container justify="center" spacing={2}>
                  {shareInformationsArray.map((share) => (
                    <Grid key={key()} item>
                      <DetailedCard
                        productName={share.productName}
                        img={share.productImg}
                        description={share.description}
                        period={share.period}
                        shippingArea={share.shippingArea}
                        days={share.days}
                        id={share.id}
                        name={share.name}
                        avatar={share.avatar}
                        comment={share.comment}
                        currentUser={currentUser}
                        rooms={rooms}
                        toggleRedirectChat={toggleRedirectChat}
                        redirectChat={redirectChat}
                        rest={rest}
                      />
                    </Grid>
                  ))}
                </Grid>
              </Grid>
            </div>
          )
      }
    </div>
  );

export default RentScreen;
