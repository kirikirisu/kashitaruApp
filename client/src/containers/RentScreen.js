import { connect } from 'react-redux';
import RentScreen from '../components/RentScreen/index';
import {
  initializeMessage,
  setRooms,
  setRoomUsers,
  setRoomName,
  setCurrentRoom,
  onMessages,
  onPresenceChanged,
  toggleRedirectChat,
  postProfile,
} from '../actions/index';

const mapStateToProps = ({ getShare, chat }) => ({
  isFetching: getShare.isFetching,
  shareInformationsArray: getShare.shareInformationsArray,
  currentUser: chat.currentUser,
  rooms: chat.rooms,
  redirectChat: getShare.redirectChat,
});

const mapDispatchToProps = (dispatch) => ({
  onMessages(message) {
    dispatch(onMessages(message));
  },
  initializeMessage() {
    dispatch(initializeMessage());
  },
  setRooms(rooms) {
    dispatch(setRooms(rooms));
  },
  setRoomUsers(roomUser) {
    dispatch(setRoomUsers(roomUser));
  },
  setRoomName(roomName) {
    dispatch(setRoomName(roomName));
  },
  setCurrentRoom(currentRoom) {
    dispatch(setCurrentRoom(currentRoom));
  },
  onPresenceChanged(currentRoom) {
    dispatch(onPresenceChanged(currentRoom));
  },
  toggleRedirectChat(bool) {
    dispatch(toggleRedirectChat(bool));
  },
  postProfile(id) {
    dispatch(postProfile(id));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(RentScreen);
