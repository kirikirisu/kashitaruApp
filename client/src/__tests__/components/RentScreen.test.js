import React from 'react';
import { shallow } from 'enzyme';
import RentScreen from '../../components/RentScreen/index';
import DetailCard from '../../components/RentScreen/DetailedCard';
import products from '../../mockApi/products';

describe('<RentScrren />', () => {
  it('<DetailCard />を表示していること', () => {
    const wrapper = shallow(<RentScreen
      isFetching={false}
      shareInformationsArray={products}
      currentUser={{}}
      rooms={{}}
      toggleRedirectChat={jest.fn()}
      redirectChat={false}
      rest={{}}
    />);

    expect(wrapper.find(DetailCard).length).toBe(7);
  });

  it('ロード画面を表示すること', () => {
    const wrapper = shallow(<RentScreen
      isFetching
      shareInformationsArray={products}
      currentUser={{}}
      rooms={{}}
      toggleRedirectChat={jest.fn()}
      redirectChat={false}
      rest={{}}
    />);

    expect(wrapper.find('h2').length).toBe(1);
  });
});
