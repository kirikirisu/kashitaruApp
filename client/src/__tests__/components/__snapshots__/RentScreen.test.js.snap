// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RentScrren /> <DetailCard />を表示していること 1`] = `
"<div>
  <div>
    <WithStyles(ForwardRef(Grid)) item={true} xs={12}>
      <WithStyles(ForwardRef(Grid)) container={true} justify=\\"center\\" spacing={2}>
        <WithStyles(ForwardRef(Grid)) item={true}>
          <DetailedCard productName=\\"カッパの尻子玉\\" img=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/Android.png?alt=media&token=aae1f59d-0da8-434a-bb0c-91808835f2e4\\" description=\\"村長のやつです\\" period=\\"８ヶ月\\" shippingArea=\\"荒川区\\" days=\\"３日\\" id={7} name=\\"カッパ\\" avatar=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/20140321122433e1d.jpg?alt=media&token=86086470-f12c-4cb6-869e-9ff8416cd887\\" comment=\\"はい妖怪幻術甲羅はずれてるように見えたの術\\" currentUser={{...}} rooms={{...}} toggleRedirectChat={[Function: mockConstructor]} redirectChat={false} rest={{...}} />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid)) item={true}>
          <DetailedCard productName=\\"カッパの皿\\" img=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/Go.png?alt=media&token=ff2fe281-1403-4f1b-9584-5501cc388e51\\" description=\\"村長のやつ\\" period=\\"９ヶ月\\" shippingArea=\\"荒川区\\" days=\\"９日\\" id={7} name=\\"カッパ\\" avatar=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/20140321122433e1d.jpg?alt=media&token=86086470-f12c-4cb6-869e-9ff8416cd887\\" comment=\\"はい妖怪幻術甲羅はずれてるように見えたの術\\" currentUser={{...}} rooms={{...}} toggleRedirectChat={[Function: mockConstructor]} redirectChat={false} rest={{...}} />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid)) item={true}>
          <DetailedCard productName=\\"fsdfsdf\\" img=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/rails2-logo.png?alt=media&token=589c6622-e31e-4499-917b-5af119b55277\\" description=\\"fsdfsdfsdf\\" period=\\"fdsff\\" shippingArea=\\"fdsf\\" days=\\"fsdf\\" id={7} name=\\"カッパ\\" avatar=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/20140321122433e1d.jpg?alt=media&token=86086470-f12c-4cb6-869e-9ff8416cd887\\" comment=\\"はい妖怪幻術甲羅はずれてるように見えたの術\\" currentUser={{...}} rooms={{...}} toggleRedirectChat={[Function: mockConstructor]} redirectChat={false} rest={{...}} />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid)) item={true}>
          <DetailedCard productName=\\"jhkhkhj\\" img=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/docker.png?alt=media&token=7fe28d4a-f2de-49e2-be42-b05914f5b692\\" description=\\"kkm\\" period=\\"thth\\" shippingArea=\\"gerg\\" days=\\"gerg\\" id={7} name=\\"カッパ\\" avatar=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/20140321122433e1d.jpg?alt=media&token=86086470-f12c-4cb6-869e-9ff8416cd887\\" comment=\\"はい妖怪幻術甲羅はずれてるように見えたの術\\" currentUser={{...}} rooms={{...}} toggleRedirectChat={[Function: mockConstructor]} redirectChat={false} rest={{...}} />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid)) item={true}>
          <DetailedCard productName=\\"星の被り物\\" img=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/Gitlab.png?alt=media&token=b9d44c7b-5467-4956-a99e-b38c21f3b0ec\\" description=\\"黄色いやつです\\" period=\\"５ヶ月\\" shippingArea=\\"荒川区\\" days=\\"２日\\" id={8} name=\\"ほし\\" avatar=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/7c47b5da.jpg?alt=media&token=c5ca5944-f442-40d5-8587-01ca148c9f17\\" comment=\\"「水着の女の子のいるところ」...それだけで十分なんだよ！！\\" currentUser={{...}} rooms={{...}} toggleRedirectChat={[Function: mockConstructor]} redirectChat={false} rest={{...}} />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid)) item={true}>
          <DetailedCard productName=\\"hogehoge\\" img=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/VScode.png?alt=media&token=e3a51472-f0f3-4234-8d22-e6975bc68ea9\\" description=\\"hogehogehogehogeghoegjeogheogheog\\" period=\\"gwregrg\\" shippingArea=\\"gererg\\" days=\\"gregerg\\" id={9} name=\\"nino\\" avatar=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/images-3-12.jpg?alt=media&token=5e6c4888-b9e6-4afb-8b82-c5c42ccd3ffc\\" comment=\\"私は金星人だ\\" currentUser={{...}} rooms={{...}} toggleRedirectChat={[Function: mockConstructor]} redirectChat={false} rest={{...}} />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid)) item={true}>
          <DetailedCard productName=\\"sfew\\" img=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/React.png?alt=media&token=0a1dc8be-29a3-4fd3-b8c7-07c5bf49bbbe\\" description=\\"fewfwe\\" period=\\"fw\\" shippingArea=\\"fwewf\\" days=\\"gweg\\" id={9} name=\\"nino\\" avatar=\\"https://firebasestorage.googleapis.com/v0/b/kashitaru-434fb.appspot.com/o/images-3-12.jpg?alt=media&token=5e6c4888-b9e6-4afb-8b82-c5c42ccd3ffc\\" comment=\\"私は金星人だ\\" currentUser={{...}} rooms={{...}} toggleRedirectChat={[Function: mockConstructor]} redirectChat={false} rest={{...}} />
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </div>
</div>"
`;

exports[`<RentScrren /> ロード画面を表示すること 1`] = `
"<div>
  <h2>
    Now Loading...
  </h2>
</div>"
`;
